drivers/storage/digitalocean/storage/digitalocean_storage.go

```diff
@@ -0,0 +1,319 @@
+}
+
+func (d *driver) VolumeAttach(ctx types.Context, volumeID string, opts *types.VolumeAttachOpts) (*types.Volume, string, error) {
+	vol, _, err := d.client.Storage.GetVolume(volumeID)
+	if err != nil {
+		return nil, "", goof.WithError("error retrieving volume", err)
+	}
+
+	if len(vol.DropletIDs) > 0 {
+		if !opts.Force {
+			return nil, "", goof.New("volume already attached!")
+		}
+
+		_, _, err := d.client.StorageActions.Detach(volumeID)
+		if err != nil {
+			return nil, "", err
+		}
+	}
+
+	dropletID := mustInstanceIDID(ctx)
+	dropletIDI, err := strconv.Atoi(*dropletID)
+	if err != nil {
+		return nil, "", err
+	}
+
+	action, _, err := d.client.StorageActions.Attach(volumeID, dropletIDI)
+	if err != nil {
+		return nil, "", err
+	}
+
+	err = d.waitForAction(volumeID, action)
+	if err != nil {
+		return nil, "", err
+	}
+
+	attachedVol, err := d.VolumeInspect(ctx, volumeID, &types.VolumeInspectOpts{
+		Attachments: types.VolAttReqTrue})
+	if err != nil {
+		return nil, "", goof.WithError("error retrieving volume", err)
+	}
+
+	return attachedVol, attachedVol.Attachments[0].DeviceName, nil
+}
+
+func (d *driver) VolumeDetach(ctx types.Context, volumeID string, opts *types.VolumeDetachOpts) (*types.Volume, error) {
+	vol, _, err := d.client.Storage.GetVolume(volumeID)
+	if err != nil {
+		return nil, goof.WithError("error getting volume", err)
+	}
+
+	if len(vol.DropletIDs) == 0 {
+		return nil, goof.WithError("volume already detached", err)
+	}
+
+	action, _, err := d.client.StorageActions.Detach(volumeID)
+	if err != nil {
+		return nil, err
+	}
+
+	err = d.waitForAction(volumeID, action)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx.Info("detached volume", volumeID)
+
+	detachedVol, err := d.VolumeInspect(ctx, volumeID, &types.VolumeInspectOpts{
+		Attachments: types.VolAttReqTrue,
+		Opts:        opts.Opts,
+	})
+	if err != nil {
+		return nil, goof.WithError("error getting volume information", err)
+	}
+
+	return detachedVol, nil
+}
+
+func (d *driver) Snapshots(ctx types.Context, opts types.Store) ([]*types.Snapshot, error) {
+	return nil, types.ErrNotImplemented
+}
+
+func (d *driver) SnapshotInspect(ctx types.Context, snapshotID string, opts types.Store) (*types.Snapshot, error) {
+	return nil, types.ErrNotImplemented
+}
+
+func (d *driver) SnapshotCopy(ctx types.Context, snapshotID string, snapshotName string, destinationID string, opts types.Store) (*types.Snapshot, error) {
+	return nil, types.ErrNotImplemented
+}
+
+func (d *driver) SnapshotRemove(ctx types.Context, snapshotID string, opts types.Store) error {
+	return types.ErrNotImplemented
+}
+
+func mustInstanceIDID(ctx types.Context) *string {
+	return &context.MustInstanceID(ctx).ID
+}
+
+func (d *driver) toTypesVolume(ctx types.Context, volume *godo.Volume, attachments types.VolumeAttachmentsTypes) *types.Volume {
+	// Collect attachment info for the volume
+	var atts []*types.VolumeAttachment
+	if attachments.Requested() {
+		for _, id := range volume.DropletIDs {
```

Comment:
This is pretty much the only way you can find out whether or not a volume is attached. If you're using `VolumeInspect` with `VolAttReqTrue` like we do in the tests, this causes it to skip adding an attachments block entirely which doesn't seem to be expected behavior. Should it default to adding a block anyway even if the volume isn't attached to a host?